docs/why-and-how-of-inline-directives.md

```diff
@@ -0,0 +1,369 @@
+
+For example, the `Right` data constructor below is "produced" when it wraps `"value`" and then immediately consumed by `eitherToMaybe`.
+
+```purs
+eitherToMaybe :: forall l r. Either l r -> Maybe r
+eitherToMaybe = case _ of
+  Left _ -> Nothing
+  Right a -> Just a
+
+eitherToMaybe $ Right "value"
+```
+
+Here's another example using records. The `{ bar: 42 }` record is "produced" by being defined in a let binding. It is then immediately "consumed" when its `bar` field is accessed. Thus, the record can be removed entirely and `a.bar` can be replaced with `42`.
+
+```purs
+foo = do
+  let
+    a = { bar: 42 }
+    b = 1 + a.bar
+  b
+```
+
+Lastly, here's an example with a few different conclusions:
+
+- If we do NOT know what `f` will do with `a` (e.g. if `f` was some FFI function), then we cannot call the record an "unnecessary intermediate data structure." `f` may need the entire record, and that possibility forces us to call the record structure necessary.
+- If we do know what `f` will do (e.g. if `f = _.bar`), then this is still an example of a primitive data "production" followed by immediate primitive data "consumption".
+
+```purs
+foo = do
+  let
+    a = { bar: 42, baz: "something" }
+    b = 1 + f a
+  b
+```
+
+**To summarize, a data producer followed immediately by a data consumer indicates a place where an unneeded intermediate data structure exists and can be removed.**
+
+### Removing Unneeded Data Structures
+
+#### Inlining and Inline Directives
+
+An inliner replaces a function's call site (e.g. stuff on the left-hand-side of the `=`) with its implementation (e.g. stuff on the right-hand-side of the `=`). This replacement means the function's body is duplicated and will appear at least twice in the resulting code: once in its original definition and once in the usage site.
+
+Here's one example using the function `binaryPlus`:
+
+```purs
+binaryPlus :: Int -> Int -> Int
+binaryPlus a b = a + b
+
+usage = 9 + binaryPlus 1 2
+```
+
+There's a few ways we could inline the expression, `binaryPlus 1 2`. While `purs-backend-es` will performance its own analysis to determine whether and when to inline it, it doesn't know as much as the developer does.
+
+A developer can use **inline directives** to tell `purs-backend-es` whether to inline a function's body to its usage site and when (i.e. how many args need to be passed to the function before the inlining occurs):
+
+- never inline it at all (i.e. the directive is `never`)
+    - `9 + binaryPlus 1 2`
+- always inline it with no consideration for the number of arguments applied to it (i.e. the directive is `always`)
+    - `9 + (\a b -> a + b) 1 2`
+- inline it only after at least one argument has been applied to it (i.e. the directive is `arity=1`)
+    - `9 + (\  b -> 1 + b)   2`
+- inline it only after at least two arguments have been applied to it (i.e. the directive is `arity=2`)
+    - `9 + (        1 + 2)`
```

Comment:
I think this example is a little confusing because there's no difference here in practice between the arities here since they call is saturated anyway. To illustrate that you have to show some partial application. That is, with `arity=2` something like `binaryPlus 1` would not inline. However with `arity=1`, you'd get `\b -> 1 + b`.